File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_59_reverse_linked_list_ii.ts
import { ListNode } from './ListNode';   

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    if (!head || left === right) {
        return head;
    }
    
    // Create a dummy node to handle edge cases where left = 1
    const dummy = new ListNode(0);
    dummy.next = head;
    
    // Step 1: Move to the node before the left position
    let prev: ListNode | null = dummy;
    for (let i = 0; i < left - 1; i++) {
        prev = prev!.next;
    }
    
    // Step 2: Reverse the sublist from left to right
    // prev points to the node before left position
    // current points to the left position
    let current: ListNode | null = prev!.next;
    
    // Reverse nodes from left to right
    for (let i = 0; i < right - left; i++) {
        // Save the next node to move
        const nextNode: ListNode | null = current!.next;
        // Point current.next to the node after nextNode
        current!.next = nextNode!.next;
        // Insert nextNode right after prev
        nextNode!.next = prev!.next;
        prev!.next = nextNode;
    }
    
    return dummy.next;
}

File: src/my_project/interviews/top_150_questions_round_22/ex_59_reverse_linked_list_ii.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head or left == right:
            return head
        
        # Create a dummy node to handle edge cases where left = 1
        dummy = ListNode(0)
        dummy.next = head
        
        # Step 1: Move to the node before the left position
        prev = dummy
        for _ in range(left - 1):
            prev = prev.next
        
        # Step 2: Reverse the sublist from left to right
        # prev points to the node before left position
        # current points to the left position
        current = prev.next
        
        # Reverse nodes from left to right
        for _ in range(right - left):
            # Save the next node to move
            next_node = current.next
            # Point current.next to the node after next_node
            current.next = next_node.next
            # Insert next_node right after prev
            next_node.next = prev.next
            prev.next = next_node
        
        return dummy.next